Add module for outputting Kraken-style taxonomy report and --taxLineage

fbreitwieser opened the pull request:
- adds module taxonomyreport for Kraken-style taxonomy report
- add parameter --taxLineage to par.lca for lca, taxonomy and addtaxonomy. Currently this means there is a variable number of fields in the output, which isn't optimal
- refactored NcbiTaxonomy:
  - does not load delnodes.dmp anymore. When a taxon that was seen in a report is not in the taxDB, it could be either through 
  - get rid of two maps; instead D is vector mapping TaxID to internal ID
  - use TaxID type to better distinguish interal IDs and taxIDs

fbreitwieser commented:
Changes in mmseqs-benchmark regression to test taxonomyreport:
```sh
# Create taxreport
"${MMSEQS}" taxonomyreport -v 3 "$TARGETDB" "$RESULTS/results_aln" "$RESULTS/results_aln_taxreport"

# Check numbers in taxreport
R_BACTERIA=$(grep 'superkingdom.*Bacteria' "$RESULTS/results_aln_taxreport" | cut -f 2)
R_VIRUS=$(grep 'superkingdom.*Virus' "$RESULTS/results_aln_taxreport" | cut -f 2)
R_EUKARYOTA=$(grep 'superkingdom.*Eukaryota' "$RESULTS/results_aln_taxreport" | cut -f 2)

TARGET="from filtertaxdb: 2524 259 2713; from taxonomyreport: 2524 259 2713"
ACTUAL="from filtertaxdb: $BACTERIA $VIRUS $EUKARYOTA; from taxonomyreport: $R_BACTERIA $R_VIRUS $R_EUKARYOTA"
awk -v actual="$ACTUAL" -v target="$TARGET" 'BEGIN { print (actual == target) ? "GOOD" : "BAD"; \
    print "Expected: ", target; \
    print "Actual:   ", actual; }' > "${RESULTS}/report"
```

milot-mirdita commented on src/commons/Parameters.cpp:
```diff
@@ -224,16 +224,16 @@ Parameters::Parameters():
+        // lca and addtaxonomy
         PARAM_LCA_RANKS(PARAM_LCA_RANKS_ID, "--lca-ranks", "LCA ranks", "Ranks to return in LCA computation", typeid(std::string), (void*) &lcaRanks, ""),
         PARAM_BLACKLIST(PARAM_BLACKLIST_ID, "--blacklist", "Blacklisted taxa", "Comma separated list of ignored taxa in LCA computation", typeid(std::string), (void*)&blacklist, "([0-9]+,)?[0-9]+"),
+        PARAM_TAXON_ADD_LINEAGE(PARAM_TAXON_ADD_LINEAGE_ID, "--taxLineage", "Add column with taxon lineage", "Show taxon lineage", typeid(bool), (void*)&showTaxLineage, ""),
```
```suggestion
        PARAM_TAXON_ADD_LINEAGE(PARAM_TAXON_ADD_LINEAGE_ID, "--tax-lineage", "Show taxon lineage", "Add column with taxon lineage", typeid(bool), (void*)&showTaxLineage, ""),
```

milot-mirdita commented on src/taxonomy/NcbiTaxonomy.cpp:
```diff
@@ -29,28 +31,36 @@ void deleteMatrix(int** M, size_t maxNodes) {
+    std::fill(H, H + maxNodes, 0);
+
+    std::vector< std::vector<TaxID> > children(taxonNodes.size());
+    for (const TaxonNode& tn : taxonNodes) {
```
Avoid C++11 constructs please

milot-mirdita commented on src/taxonomy/NcbiTaxonomy.cpp:
```diff
@@ -104,89 +114,45 @@ std::vector<std::string> splitByDelimiter(const std::string &s, const std::strin
+        if (taxId > maxTaxID) {
+            maxTaxID = taxId;
         }
+        taxonNodes.emplace_back(TaxonNode(currentId, taxId, parentTaxId, result[2]));
```
```suggestion
        taxonNodes.emplace_back(currentId, taxId, parentTaxId, result[2]);
```
Let it automatically forward the parameters with emplace_back

milot-mirdita commented on src/taxonomy/NcbiTaxonomy.cpp:
```diff
@@ -104,89 +114,45 @@ std::vector<std::string> splitByDelimiter(const std::string &s, const std::strin
-        for (size_t i = 0; i < node.children.size(); ++i) {
-            node.children[i] = D[node.children[i]];
+    // Loop over taxonNodes and check no parents are NULL
+    for (const TaxonNode& tn : taxonNodes) {
```
Same as above, I would avoid the : for loop

milot-mirdita commented on src/taxonomy/NcbiTaxonomy.h:
```diff
@@ -6,55 +6,63 @@
 #include <vector>
 #include <string>
 
+using TaxID = int;
```
Can we do a typedef instead of using?

fbreitwieser replied:
Sure. Have a couple of C++11-isms, the others being the foreach loop as you noted. Martin said that he'll discuss them with you whether you want to have them, but as long as it's not auto he is not opposed.

martin-steinegger replied:
I noticed that we use 0x for the GNU compiler. I am not sure why C++-11 is active for llvm. We should avoid the foreach loops. Sorry for the confusion.

fbreitwieser replied:
Oh ok, not too much to change. Will fix it.

fbreitwieser commented:
> Hi Florian,
> 
> Thanks a lot for your pull request. I added some comments. One thing that I am unsure about is the removal of delnodes however. I added that explicitly since an old Uniclust release can have taxids that are already removed in the latest NCBI taxonomy. Then we either get an error or we ignore the protein in Uniprot/Uniclust.

Hi Milot, thansk for looking through it! Re deleted nodes: The reason I think they are not necessary is that for mmseqs it does not matter whether a taxon that is encountered in an report, but is not in the taxonomy database, is from a deleted taxon or a taxon that has been added after the creation of the taxonomy database. In both cases the taxon can be ignored (or an error can be thrown).

martin-steinegger commented:
I think we can leave out the deleted nodes. It makes the code also more clean.

fbreitwieser commented:
Btw, what's your policy on double right-angle brackets? E.g. `std::vector<std::vector<int>>`. Wih space or not? Was a problem pre C++11, but maybe works in 0x?
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2005/n1757.html

martin-steinegger commented:
The regression test finished without errors, however the binary could not be uploaded. Maybe its a problem with the PR. I merged the commit into the master.  
So MMseqs2 is one step closer to be a great taxonomy classification tool. Thanks a lot!
